### PR TITLE
[FLUME-3100] Support arbitrary header substitution for topic of Kafka

### PR DESCRIPTION
**Pull request by stakafum**

### CODE REVIEW

File: flume-ng-sinks/flume-ng-kafka-sink/src/test/java/org/apache/flume/sink/kafka/TestKafkaSink.java [hunk 1]
```diff
@@ -239,6 +240,42 @@ public void testTopicAndKeyFromHeader() throws UnsupportedEncodingException {
...
+    Context context = prepareDefaultContext();
+    context.put(TOPIC_CONFIG, TestConstants.HEADER_TOPIC);
+    Configurables.configure(kafkaSink, context);
+    Channel memoryChannel = new MemoryChannel();
+    Configurables.configure(memoryChannel, context);
+    kafkaSink.setChannel(memoryChannel);
+    kafkaSink.start();
+
+    String msg = "test-replace-substring-of-topic-with-headers";
+    Map<String, String> headers = new HashMap<String, String>();
```
Comment (adenes) [inline flume-ng-sinks/flume-ng-kafka-sink/src/test/java/org/apache/flume/sink/kafka/TestKafkaSink.java]: Nit: explicit type argument can be replaced by `<>`. Or `Collections.singletonMap()` might be even better.

**Review by adenes — CHANGES_REQUESTED** [review]
Thank you for the patch @stakafum, it's a great enhancement to the Kafka sink.
I had one minor comment, otherwise it looks good to me.

**Comment by stakafum** [comment]
@adenes  Thanks for the review. I have updated my pull request.

**Comment by adenes** [comment]
Thank you for the updated patch, it looks good to me, +1
I'm about to commit this.

(Did you amend the commit and force-pushed? For a more complex pull request it might be better to have multiple commits to be able to see the differences between them. The committers will squash the commits before pushing to the main branch.)

**Pull request closed without merging**